File: 45-jump-game-ii/jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int res = 0, i = 0, size = nums.size();
        if (size == 0 || size == 1) return 0;
        if (size == 2 && nums[0]!=0) return 1;

        while (i < size){
            if (i+nums[i]>= size-1) break;
            priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> temp;
            int jmp = nums[i];
            for(int j = i+1; j<=i+jmp; j++) temp.emplace(size - nums[j] - j, j);
            auto t = temp.top();
            i = t.second;
            res++;
        }

        return res+1;
    }
};

File: 45-jump-game-ii/jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        ln = len(nums)

        if ln == 1:
            return 0

        list_jump = deque()
        curr_jump = (nums[0], 1)

        for i in range(1, ln):
            if curr_jump[0]<i+nums[i]:
                list_jump.append((i+nums[i], curr_jump[1]+1))
            if curr_jump[0] >= ln-1:
                return curr_jump[1]
            
            while list_jump and curr_jump[0] == i:
                curr_jump = max(curr_jump, list_jump.popleft())
        return curr_jump[1]
        